### DISCUSSION TITLE
Saving Data in an Array on Crazyflie via Python Script

### DISCUSSION THREAD

**Discussion by amitx64** [bitcraze/discussions#1263] [Q&A]
Hi @tobbeanton and @ataffanel,

I'm interested in delving into the process of saving data in an array on Crazyflie using a Python script that utilizes MEM_TYPE_APP. The goal is to eventually expand this functionality to save data on an SD card once the array storage method proves successful.

Do you happen to have any examples that I could use to experiment with this?

**Comment by ataffanel** [comment]
No, I am not aware of any example doing this.

**Reply by amitx64** [reply]
Oh! So, just tell me how I can `send` some `data to Crazyflie` via a `Python script`. After receiving that data, Crazyflie stores it in an array utilizing `MEM_TYPE_APP` and `memHandler`. I'm asking because `Bitcraze` does the same thing to upload `trajectories` onto `Trajectory memory` via a Python script in `autosequence_commander_high_level.py`, utilizing `MEM_TYPE_TRAJ`. How can I achieve the same functionality with `MEM_TYPE_APP`? Help me out with this.

**Reply by ataffanel** [reply]
The API is going to be the same whatever memory is accessed, both in Python and C. So if you look at how it is done for another memory type, this would be a good example of how to do it with the app memory.

**Reply by amitx64** [reply]
Hey @ataffanel, I'm reaching out because I noticed that in the `crazyflie-lib-python` repository, there's a `mem` folder containing `trajectory_memory.py`, but I couldn't find `app_memory.py`. This is leaving me a bit confused. Do you think it's necessary to add the `app_memory.py` file there? I'm thinking it might be crucial for sending data to the `app memory` in `crazyflie`.

**Reply by ataffanel** [reply]
The app memory is intended to be used by scripts for custom purpose. The idea was that one would directly use the memory api to communicate with an app. An app_memory.py abstacting this and making it easier to use might be a useful thing and we would gladly accept a PR implementing it. However I do not think it is "capital" in the sense that scripts can directly use the memory API with the right parameters.

**Reply by amitx64** [reply]
Hi @ataffanel,

I'm asking the same question repeatedly because I noticed something interesting in the `crazyflie-firmware` modified by `skybrush`. They're storing light program data in the `lightProgramMemory` array, and I'm puzzled by how they accomplish this in `light_program.c`. Here's the code:

```c
.....
....
/**
 * Size of the memory segment that can store light programs.
 */
#define LIGHT_PROGRAM_MEMORY_SIZE 4096
#define NUM_LIGHT_PROGRAM_DEFINITIONS 10
uint8_t lightProgramMemory[LIGHT_PROGRAM_MEMORY_SIZE];
.....
.....
// Light program memory handling from the memory module
static uint32_t handleMemGetSize(void) { return lightProgramPlayerMemSize(); }
static bool handleMemRead(const uint32_t memAddr, const uint8_t readLen, uint8_t* buffer);
static bool handleMemWrite(const uint32_t memAddr, const uint8_t writeLen, const uint8_t* buffer);
static const MemoryHandlerDef_t memDef = {
  .type = MEM_TYPE_APP,
  .getSize = handleMemGetSize,
  .read = handleMemRead,
  .write = handleMemWrite,
};
.....
....
static bool handleMemRead(const uint32_t memAddr, const uint8_t readLen, uint8_t* buffer) {
  bool result = false;

  if (memAddr + readLen <= sizeof(lightProgramMemory)) {
    memcpy(buffer, &(lightProgramMemory[memAddr]), readLen);
    result = true;
  }

  return result;
}

static bool handleMemWrite(const uint32_t memAddr, const uint8_t writeLen, const uint8_t* buffer) {
  bool result = false;

  if ((memAddr + writeLen) <= sizeof(lightProgramMemory)) {
    memcpy(&(lightProgramMemory[memAddr]), buffer, writeLen);
    result = true;
  }

  return result;
}
.....
.....
```

I'm confused because, as you can see in the code snippet, they're using `MEM_TYPE_APP` to store some data in the `lightProgramMemory` array and utilizing `MemoryHandlerDef_t`. Could you assist me in achieving a similar functionality. So, I can test that using Python.

**Reply by ataffanel** [reply]
Yes I understand what you want to do, I do not understand what the problem is though. The python library has all necessary APIs to communicate with different kind of memories in the Crazyflie, where exactly are you getting stuck. What API call is not working for you?

**Reply by amitx64** [reply]
@ataffanel, Thank you. I just figured out that part in Python. Actually, due to a mistake in my Python script, I always got the memory type as unknown. I added a few lines of code in cflib, and that resolved my issues. Thanks to your support, I was able to identify the issue in my Python script.
